scripts/machine_translation/train_gnmt.py

```diff
@@ -195,48 +217,58 @@ def train():
     trainer = gluon.Trainer(model.collect_params(), args.optimizer, {'learning_rate': args.lr})
 
     train_data_loader, val_data_loader, test_data_loader \
-        = dataprocessor.make_dataloader(data_train, data_val, data_test, args)
+        = dataprocessor.make_dataloader(data_train, data_val, data_test,
+                                        args, num_shards=len(context))
 
     best_valid_bleu = 0.0
     for epoch_id in range(args.epochs):
         log_avg_loss = 0
-        log_avg_gnorm = 0
         log_wc = 0
+        loss_denom = 0
+        step_loss = 0
+
         log_start_time = time.time()
-        for batch_id, (src_seq, tgt_seq, src_valid_length, tgt_valid_length)\
+        for batch_id, seqs \
                 in enumerate(train_data_loader):
-            # logging.info(src_seq.context) Context suddenly becomes GPU.
-            src_seq = src_seq.as_in_context(ctx)
-            tgt_seq = tgt_seq.as_in_context(ctx)
-            src_valid_length = src_valid_length.as_in_context(ctx)
-            tgt_valid_length = tgt_valid_length.as_in_context(ctx)
+
+            src_wc, tgt_wc, bs = np.sum([(shard[2].sum(), shard[3].sum(), shard[0].shape[0])
+                                         for shard in seqs], axis=0)
+            src_wc = src_wc.asscalar()
+            tgt_wc = tgt_wc.asscalar()
+
+            loss_denom += tgt_wc - bs
+            seqs = [[seq.as_in_context(_context) for seq in shard]
+                    for _context, shard in zip(context, seqs)]
+
+            LS = []
             with mx.autograd.record():
-                out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
-                loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
-                loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
-                loss.backward()
-            grads = [p.grad(ctx) for p in model.collect_params().values()]
-            gnorm = gluon.utils.clip_global_norm(grads, args.clip)
-            trainer.step(1)
-            src_wc = src_valid_length.sum().asscalar()
-            tgt_wc = (tgt_valid_length - 1).sum().asscalar()
-            step_loss = loss.asscalar()
-            log_avg_loss += step_loss
-            log_avg_gnorm += gnorm
+                for src_seq, tgt_seq, src_valid_length, tgt_valid_length in seqs:
+                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
+
+                    ls = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).sum()
+                    LS.append(ls * (tgt_seq.shape[1] - 1))
+            for L in LS:
+                L.backward()
+
+            trainer.step(float(loss_denom))
```

Comment:
We need to divide `loss_denom()` before calling `step()`, i.e, always divide first and call `step(1.0)`. This ensures that the head_gradient will not be too large. If the head gradient have a large scale, the calculated gradient will also have a large scale and may explode in the backpropagation process.

Reply:
Thanks for review, I see the transformer do the step() with `loss_denom()`, and i think step() function will make loss averaged by the input parameter(loss_denorm), so the gradient will not be too large, am i right?

- transformer code

https://github.com/dmlc/gluon-nlp/blob/1a5f9f38c9f2d070a8f6af6a8234d8e1daed5778/scripts/machine_translation/train_transformer.py#L321

- source code of step()

https://github.com/apache/incubator-mxnet/blob/d2102faa228bdc6723a9da299c6ff5999cbbdcdb/python/mxnet/gluon/trainer.py#L276

Reply:
If the loss is too large, the step `loss.backward()` can explode and lead to NAN. That's why we rescaled the loss.